File: Leetcode_POTD_2025/10_OCTOBER/1625_Lexicographically_Smallest_String_After_Applying_Operations.cpp
#include <string>
#include <queue>
#include <unordered_set>
#include <algorithm>
#include <iostream>

using namespace std;

class Solution {
private:
    string flux_add(string s_in, int a_val) {
        string s_out = s_in;
        for (size_t i = 1; i < s_out.length(); i += 2) {
            s_out[i] = ((s_out[i] - '0' + a_val) % 10) + '0';
        }
        return s_out;
    }

    string cycle_rotate(string s_in, int b_val) {
        int len = s_in.length();
        b_val %= len;
        if (b_val == 0) return s_in;
        return s_in.substr(len - b_val) + s_in.substr(0, len - b_val);
    }

public:
    string findLexSmallestString(string s, int a, int b) {
        unordered_set<string> temporal_set;
        queue<string> path_queue;
        string chronos_key = s;

        path_queue.push(s);
        temporal_set.insert(s);

        while (!path_queue.empty()) {
            string current_moment = path_queue.front();
            path_queue.pop();
            
            if (current_moment < chronos_key) {
                chronos_key = current_moment;
            }

            string next_add = flux_add(current_moment, a);
            if (temporal_set.find(next_add) == temporal_set.end()) {
                temporal_set.insert(next_add);
                path_queue.push(next_add);
            }

            string next_rot = cycle_rotate(current_moment, b);
            if (temporal_set.find(next_rot) == temporal_set.end()) {
                temporal_set.insert(next_rot);
                path_queue.push(next_rot);
            }
        }
        return chronos_key;
    }
};

File: Leetcode_POTD_2025/10_OCTOBER/1625_Lexicographically_Smallest_String_After_Applying_Operations.java
import java.util.ArrayDeque;
import java.util.HashSet;
import java.util.Queue;
import java.util.Set;

class Solution {
    private String apply_add_op(String s_in, int a_val) {
        char[] arr = s_in.toCharArray();
        for (int i = 1; i < arr.length; i += 2) {
            arr[i] = (char) (((arr[i] - '0' + a_val) % 10) + '0');
        }
        return new String(arr);
    }

    private String apply_rot_op(String s_in, int b_val) {
        int len = s_in.length();
        b_val %= len;
        if (b_val == 0) return s_in;
        return s_in.substring(len - b_val) + s_in.substring(0, len - b_val);
    }

    public String findLexSmallestString(String s, int a, int b) {
        Set<String> nebula_set = new HashSet<>();
        Queue<String> path_queue = new ArrayDeque<>();
        String stellar_core = s;

        path_queue.offer(s);
        nebula_set.add(s);

        while (!path_queue.isEmpty()) {
            String current_star = path_queue.poll();
            
            if (current_star.compareTo(stellar_core) < 0) {
                stellar_core = current_star;
            }

            String next_add = apply_add_op(current_star, a);
            if (!nebula_set.contains(next_add)) {
                nebula_set.add(next_add);
                path_queue.offer(next_add);
            }

            String next_rot = apply_rot_op(current_star, b);
            if (!nebula_set.contains(next_rot)) {
                nebula_set.add(next_rot);
                path_queue.offer(next_rot);
            }
        }
        return stellar_core;
    }
}
